### PR TITLE
chore(feg): Bump go-control-plane version

### PR DESCRIPTION
**Pull request by alexzurbonsen**
Prerequisite for https://github.com/magma/magma/pull/12151.

## Summary

To prepare the larger golang update, this PR updates the module `github.com/envoyproxy/go-control-plane` to `v.0.10.1` in `feg/gateway`.

Why is this necessary?

1) To make the golang update work we want to update `github.com/go-swagger/go-swagger` to `v0.29.0` (latest available version) in `orc8r/cloud/go` module. It has `github.com/envoyproxy/go-control-plane` at `v0.10.1` as an indirect dependency.
Since the `feg/gateway` module depends on the `orc8r/cloud/go` module, Go's minimal version selection algorithm forces us to use `github.com/envoyproxy/go-control-plane` at `v0.10.1`. But, at least for packages `cache` and `server`, `v0.10.1` is not backwards compatible with version `v0.9.4`. (Go expects that they are, since they have the same major version.)

2) Also versions of `google.golang.org/grpc` that are `>= v1.34.0` (latest: `v1.45.0`) depend on `github.com/envoyproxy/go-control-plane > v0.9.4`. This versions aren't fully backwards compatible either. An incompatible version of `google.golang.org/grpc` is likely to appear in the dependency graph at some point.
Consequently, using a lower version of  `github.com/go-swagger/go-swagger`, like `v0.27.0`, which does not depend on a version `> v0.9.4` of `github.com/envoyproxy/go-control-plane`, does not permanently solve this. 

## Test Plan

`make test` in feg/gateway passes. (Ran `make precommit` in `feg/gateway`)

## Additional Information

- Done in pairing with @voisey and @ajahl.

### CODE REVIEW

**Comment by alexzurbonsen** [comment]
@maxhbr Can you have another look at the license compliance related changes? Thanks!

File: feg/gateway/services/envoy_controller/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+This directory contains Magma's Envoy Controller Service. It configures Envoy proxy with specified configuration.
+
+It uses code from github.com/envoyproxy/go-control-plane module which is published under copyright of Envoyproxy Authors
+and license Apache 2.0. Please find the license text in envoy_controller/LICENSE.Apache-2.0.
```
Comment (Neudrino) [inline feg/gateway/services/envoy_controller/README.md]: You could make this a link to the correct file.
Something along the lines
```suggestion
and [Apache 2.0 license](LICENSE.Apache-2.0).
```

File: feg/gateway/services/envoy_controller/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
```
Comment (Neudrino) [inline feg/gateway/services/envoy_controller/README.md]: EOL at EOF is fine, but I think we do not need a empty line in addition.

**Comment by maxhbr** [comment]
License compliance looks good. Maybe create an ticket to use that project as a library to get rid of copied code

**Review by maxhbr — COMMENTED** [review]
License compliance looks good

File: feg/gateway/services/envoy_controller/control_plane/control_plane.go [hunk 3]
```diff
@@ -10,15 +44,21 @@ import (
...
+	hcm "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	clusterservice "github.com/envoyproxy/go-control-plane/envoy/service/cluster/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	endpointservice "github.com/envoyproxy/go-control-plane/envoy/service/endpoint/v3"
+	listenerservice "github.com/envoyproxy/go-control-plane/envoy/service/listener/v3"
+	routeservice "github.com/envoyproxy/go-control-plane/envoy/service/route/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	xds "github.com/envoyproxy/go-control-plane/pkg/server/v3"
```
Comment (Neudrino) [inline feg/gateway/services/envoy_controller/control_plane/control_plane.go:61]: I do not see, if those entries are ordered in some way.
If so how?
If not, can we do so, please. I suggest a lexicographical ordering.

File: feg/gateway/services/envoy_controller/control_plane/control_plane.go [hunk 4]
```diff
@@ -121,8 +165,33 @@ func (cb *callbacks) OnFetchRequest(ctx context.Context, req *v2.DiscoveryReques
...
-func (cb *callbacks) OnFetchResponse(*v2.DiscoveryRequest, *v2.DiscoveryResponse) {
-	glog.Infof("OnFetchResponse...")
+func (cb *callbacks) OnFetchResponse(*discovery.DiscoveryRequest, *discovery.DiscoveryResponse) {
+	glog.V(2).Infof("OnFetchResponse...")
+}
+
+func (cb *callbacks) OnDeltaStreamOpen(_ context.Context, id int64, typ string) error {
+	glog.V(2).Infof("delta stream %d open for %s\n", id, typ)
+	return nil
+}
```
Comment (Neudrino) [inline feg/gateway/services/envoy_controller/control_plane/control_plane.go:175]: Spacing seems to be inconsistent here. Also below.

File: feg/gateway/services/envoy_controller/control_plane/control_plane.go [hunk 5]
```diff
@@ -121,8 +165,33 @@ func (cb *callbacks) OnFetchRequest(ctx context.Context, req *v2.DiscoveryReques
...
+func (cb *callbacks) OnStreamDeltaRequest(int64, *discovery.DeltaDiscoveryRequest) error {
+	glog.V(2).Infof("OnStreamDeltaRequest...")
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+	cb.deltaRequests++
+	if cb.signal != nil {
+		close(cb.signal)
+		cb.signal = nil
+	}
+	return nil
```
Comment (Neudrino) [inline feg/gateway/services/envoy_controller/control_plane/control_plane.go:197]: _Question:_ Could you say why this is necessary. Not sure, I understand what happens here. :thinking:

File: feg/gateway/services/envoy_controller/control_plane/control_plane.go (hunk 3 above)
Comment (alexzurbonsen) [reply feg/gateway/services/envoy_controller/control_plane/control_plane.go:61]: I was also confused by this, but it turns out that they are ordered according to the module path, e.g. "github.com/...", not by the alias. So they are actually considered sorted.

Comment (voisey) [reply feg/gateway/services/envoy_controller/control_plane/control_plane.go:61]: Yes, they were sorted automatically for me as I saved my changes

File: feg/gateway/services/envoy_controller/control_plane/control_plane.go (hunk 5 above)
Comment (alexzurbonsen) [reply feg/gateway/services/envoy_controller/control_plane/control_plane.go:197]: This, and the other `On...` methods are not called in the code. They are just included to make the `callbacks` type satisfy an interface (`xds.Callbacks`).

Our contribution is to add the four functions that have Delta in the name to the set of `On...` functions. The ones that already existed seem to come from `pkg/test/v3/callbacks.go` (or a predecessor), with some modifications added. We followed that schema to be consistent.

**Review by Neudrino — APPROVED** [review]

**Review by voisey — APPROVED** [review]

**Review by maxhbr — APPROVED** [review]
LGTM

**Review by uri200 — APPROVED** [review]
Looks good to me, but @koolzz  may need to take a look at this to confirm

**Comment by koolzz** [comment]
Can you make sure the s1 he tests pass?
[test_attach_detach_rar_tcp_he.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_rar_tcp_he.py)
and
[test_attach_detach_with_he_policy.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_he_policy.py)

**Comment by alexzurbonsen** [comment]
> Can you make sure the s1 he tests pass? [test_attach_detach_rar_tcp_he.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_rar_tcp_he.py) and [test_attach_detach_with_he_policy.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_he_policy.py)

@koolzz Looked into the two integ tests. It seems they are broken since almost a year. They are commented out in `lte/gateway/python/integ_tests/defs.mk` as flaky or not working in CI. `test_attach_detach_with_he_policy.py` is supposed to run locally, according to the comment there. Tried to run it but it failed on master (executed it as part of the extended tests).

There are also some issues that tracked this tests:
- #6254
- #6439

Since issues with the two tests haven't been fixed, see this [slack communication](https://magmacore.slack.com/archives/C02HRAXJEBB/p1648656061304969). However this tests should either be fixed or removed. Fixing is out of scope in this PR.

To proceed we suggest to merge anyway, as all unit test seem to be stable and CI passes as well.

**Comment by alexzurbonsen** [comment]
@koolzz Is there any update on this one?

**Comment by koolzz** [comment]
> > Can you make sure the s1 he tests pass? [test_attach_detach_rar_tcp_he.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_rar_tcp_he.py) and [test_attach_detach_with_he_policy.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_he_policy.py)
> 
> @koolzz Looked into the two integ tests. It seems they are broken since almost a year. They are commented out in `lte/gateway/python/integ_tests/defs.mk` as flaky or not working in CI. `test_attach_detach_with_he_policy.py` is supposed to run locally, according to the comment there. Tried to run it but it failed on master (executed it as part of the extended tests).
> 
> There are also some issues that tracked this tests:
> 
> * [[s1aptest] TestAttachDetachRarTcpDataWithHE is flaky / or broken #6254](https://github.com/magma/magma/issues/6254)
> * [[s1aptest] test_attach_detach_with_he_policy.py is broken on CI  #6439](https://github.com/magma/magma/issues/6439)
> 
> Since issues with the two tests haven't been fixed, see this [slack communication](https://magmacore.slack.com/archives/C02HRAXJEBB/p1648656061304969). However this tests should either be fixed or removed. Fixing is out of scope in this PR.
> 
> To proceed we suggest to merge anyway, as all unit test seem to be stable and CI passes as well.

@vikramgreddy Do you have any updates on the 2 issues?

**Comment by vikramgreddy** [comment]
> > > Can you make sure the s1 he tests pass? [test_attach_detach_rar_tcp_he.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_rar_tcp_he.py) and [test_attach_detach_with_he_policy.py](https://github.com/magma/magma/blob/master/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_with_he_policy.py)
> > 
> > 
> > @koolzz Looked into the two integ tests. It seems they are broken since almost a year. They are commented out in `lte/gateway/python/integ_tests/defs.mk` as flaky or not working in CI. `test_attach_detach_with_he_policy.py` is supposed to run locally, according to the comment there. Tried to run it but it failed on master (executed it as part of the extended tests).
> > There are also some issues that tracked this tests:
> > 
> > * [[s1aptest] TestAttachDetachRarTcpDataWithHE is flaky / or broken #6254](https://github.com/magma/magma/issues/6254)
> > * [[s1aptest] test_attach_detach_with_he_policy.py is broken on CI  #6439](https://github.com/magma/magma/issues/6439)
> > 
> > Since issues with the two tests haven't been fixed, see this [slack communication](https://magmacore.slack.com/archives/C02HRAXJEBB/p1648656061304969). However this tests should either be fixed or removed. Fixing is out of scope in this PR.
> > To proceed we suggest to merge anyway, as all unit test seem to be stable and CI passes as well.
> 
> @vikramgreddy Do you have any updates on the 2 issues?

@ulaskozat, It was falling because of the PR: https://github.com/magma/magma/pull/11101.
Alex has reverted the commit it should work now (https://github.com/magma/magma/pull/12390)

**Comment by alexzurbonsen** [comment]
Do we expect this to fix the HE tests as well? Just wondering, since #11101 is four days old whereas the two HE tests seem to be broken since about a year ago.  @vikramgreddy

**Comment by uri200** [comment]
Hey @alexzurbonsen  you may need to undo this PR 

https://github.com/magma/magma/pull/12386

This was just added to prevent envoy to go up automatically. So maybe you need to remove that module from go.mod

**Comment by koolzz** [comment]
Can we test those s1 tests again then?

**Comment by alexzurbonsen** [comment]
> Hey @alexzurbonsen you may need to undo this PR
> 
> #12386
> 
> This was just added to prevent envoy to go up automatically. So maybe you need to remove that module from go.mod

Thanks for the hint @uri200. Rebased and removed the comment in the `google.golang.org/genproto` line.

**Comment by uri200** [comment]
@alexzurbonsen Can you try to remove the whole genproto line and see if it is needed at all? Run ./build.py -c to make sure it works.

The reason to add that line was to avoid envoy to go over 0.9.4. now that we have updated that we may not even need that line.

Maybe when you remove it, go will add it automatically. Then it is fine to keep it.

**Comment by alexzurbonsen** [comment]
> Can we test those s1 tests again then?

@koolzz Looking into it. Both seem to work on master again.

**Comment by alexzurbonsen** [comment]
> @alexzurbonsen Can you try to remove the whole genproto line and see if it is needed at all? Run ./build.py -c to make sure it works.
> 
> The reason to add that line was to avoid envoy to go over 0.9.4. now that we have updated that we may not even need that line.
> 
> Maybe when you remove it, go will add it automatically. Then it is fine to keep it.

@uri200 Yes, left it in there because I saw that [feg/gateway/services/testcore/hss/storage/storage_test.go](https://github.com/magma/magma/blob/master/feg/gateway/services/testcore/hss/storage/storage_test.go) depends on it, i.e. `go mod tidy` produces the same line after deleting it.

On the other side, I tried to delete it and run `./build.py -c` just now. Passes without errors... (?)

**Comment by uri200** [comment]
sounds good then

**Comment by alexzurbonsen** [comment]
Thanks for merging @uri200 :)

**Pull request merged**